src/main/scala/akka/persistence/jdbc/query/scaladsl/JdbcReadJournal.scala

```diff
@@ -139,46 +141,87 @@ class JdbcReadJournal(config: Config)(implicit val system: ExtendedActorSystem)
-    *                             will be returned completely.
-    */
+   * @param terminateAfterOffset If None, the stream never completes. If a Some, then the stream will complete once a
+   *                             query has been executed which might return an event with this offset (or a higher offset).
+   *                             The stream may include offsets higher than the value in terminateAfterOffset, since the last batch
+   *                             will be returned completely.
+   */
   private def eventsByTag(tag: String, offset: Long, terminateAfterOffset: Option[Long]): Source[EventEnvelope, NotUsed] = {
+
     import akka.pattern.ask
     implicit val askTimeout: Timeout = Timeout(readJournalConfig.journalSequenceRetrievalConfiguration.askTimeout)
-    /* We unfold with 3 parameters:
-     * - The first parameter (from) determines the starting offset (i.e. the minimum value) to query from.
-     * - The second parameter (completeImmediately) signifies that we have queried events until at least
-     *   `terminateAfterOffset`, therefore the stream no longer needs to continue.
-     * - The third parameter (retrieveImmediately) is a boolean which determines whether the database query should be
-     *   executed immediately, or after a delay.
+
+    sealed trait FlowControl
+
+    /** Keep querying - used when we are sure that there is more events to fetch */
+    object Continue extends FlowControl
+
+    /**
+     * Keep querying with delay - used when we have consumed all events,
+     * but want to poll for future events
      */
-    Source.unfoldAsync[(Long, Boolean, Boolean), Seq[EventEnvelope]]((offset, false, false)) {
-      case (from: Long, completeImmediately: Boolean, retrieveImmediately: Boolean) =>
-        def retrieveNextBatch(): Future[Some[((Long, Boolean, Boolean), Seq[EventEnvelope])]] = {
+    object ContinueDelayed extends FlowControl
+
+    /** Stop querying - used when we reach the desired offset  */
+    object Stop extends FlowControl
+
+    Source.unfoldAsync[(Long, FlowControl), Seq[EventEnvelope]]((offset, Continue)) {
+
+      case (from, control) =>
+
+        val batchSize = readJournalConfig.maxBufferSize
+
+        def retrieveNextBatch() = {
           for {
             queryUntil <- journalSequenceActor.ask(GetMaxOrderingId).mapTo[MaxOrderingId]
-            xs <- currentJournalEventsByTag(tag, from, readJournalConfig.maxBufferSize, queryUntil).runWith(Sink.seq)
+            xs <- currentJournalEventsByTag(tag, from, batchSize, queryUntil).runWith(Sink.seq)
           } yield {
-            val isFullBatch = xs.size == readJournalConfig.maxBufferSize
-            val completeStreamAfterThisBatch = !isFullBatch && terminateAfterOffset.exists(_ <= queryUntil.maxOrdering)
+
+            // calculates if we have more events to fetch
+            val hasMoreEvents = {
```

Comment:
We can simplify this to `val hasMoreEvents = xs.size == batchSize`

The only slight difference is the case where the possible size range is equal to batchSize. However I personally favour simple and understandable code above complex codes that handles every edge case with the best efficiency.

Reply:
I will move back to simply `xs.size == bacthSize`. 

My point was not to cover all edge cases, but to make it understandable. For instance, `val hasMoreEvents = xs.size == batchSize` doesn't convey the mental work we did to come to the conclusion that there is more events, or not. 

But I agree that my trial is not doing it better neither.